source-map.bs

```diff
@@ -590,6 +611,65 @@ as per [=ECMAScript Lexical Grammar=]. [=Mapping=] entries must point to either:
     [=Comment=], [=HashbangComment=], [=StringLiteral=], [=Template=],
     [=TemplateSubstitutionTail=], [=WhiteSpace=] and [=LineTerminator=].
 
+### [=decoded mapping/Names=] for generated JavaScript code
+
+Source map generators should create a [=decoded mapping|mapping=] entry with a
+[=decoded mapping/name=] field for a JavaScript token, if
+
+1. The original source language construct maps semantically to the generated
+    JavaScript code.
+
+1. The original source language construct has a name.
+
+Then the [=decoded mapping/name=] of the [=decoded mapping|mapping=] entry should be
+the name of the original source language construct. A [=decoded mapping|mapping=] with
+a non-null [=decoded mapping/name=] is called a <dfn>named mapping</dfn>.
+
+Example: A TypeScript function compiled to a JavaScript arrow function, or a
+local variable that was renamed.
+
+The following enumeration lists productions of the ECMAScript [=Syntactic Grammar=]
+and the respective token or non-terminal (on the right-hand side of the production)
+for which source map generators should emit a [=named mapping=].
+The [=decoded mapping|mapping=] entry created for such tokens must follow [[#mappings-javascript]].
+
+The enumeration should be understood as the "minimum". In general,
+source map generators are free to emit any additional [=named mappings=].
+
+Note: The enumeration also lists tokens where generators "may" emit named [=decoded mapping|mappings=] in addition
+to the tokens where they "should". These reflect the reality where existing tooling emits or
+expects [=named mappings=]. The duplicated [=named mapping=] is comparably cheap: Indices into
+[=json/names=] are encoded relative to each other so subsequent mappings to the same name are
+encoded as 0 (`A`).
+
+1. The [=BindingIdentifier=](s) for [=LexicalDeclaration=], [=VariableStatement=]
+    and [=Parameter List=].
+
+1. The [=BindingIdentifier=] for [=FunctionDeclaration=], [=FunctionExpression=],
+    [=AsyncFunctionDeclaration=], [=AsyncFunctionExpression=], [=GeneratorDeclaration=],
+    [=GeneratorExpression=], [=AsyncGeneratorDeclaration=], and [=AsyncGeneratorExpression=] if it
+    exists, or the opening parenthesis `(` preceding the [=FormalParameters=] otherwise.
+
+    Source map generators may chose to emit a [=named mapping=] on the opening parenthesis regardless of
+    the presence of the [=BindingIdentifier=].
+
+1. The [=BindingIdentifier=] for [=ArrowFunction=], and [=AsyncArrowFunction=] if it
```

Comment:
This paragraph is confusing and doesn't properly address the concern I raised. It now says to emit a name mapping on (a) either the _BindingIdentifier_ or `(` of an arrow function, or (b) the `=>` if it is a single argument arrow function with neither a _BindingIdentifier_ nor _Arguments_.

(b) is never true as a single argument arrow function has _only_ a _BindingIdentifier_, and thus (a) is only the ever case that is true.

Correct me if I'm wrong, but I believe we discussed in the meeting that there are two possible places to emit a name mapping for an arrow function or async arrow function:

1. For an _ArrowFunction_ or _AsyncArrowFunction_ with _Arguments_, the name mapping is attached to the opening `(`. A generator _may_ opt to emit a name mapping attached to the `=>` for consistency with (2).
2. For an _ArrowFunction_ or _AsyncArrowFunction_ with a _SingleNameBinding_, the name mapping is attached to the `=>`.

Reply:
I'm embarrassed to say that I don't know what I was thinking when I wrote this.

I fixed it up as proposed but with concrete references to the productions.